tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,282 @@
+        in_name = arg_name + "_list_in"
+        out_name = arg_name + "_list_out"
+        elem_name = arg_name + "_elem"
+        code = [f"const c10::List<c10::IValue> {in_name} = {ival}.toList();"]
+        res = argumenttype_ivalue_convert(t.elem, elem_name, arg_name)
+        connector = "\n\t"
+        # handle list type with size, e.g., bool[4]
+        if isinstance(t.elem, BaseType) and t.elem.name == BaseTy.bool and t.size:
+            ctype = f"std::array<{res.ctype}, {t.size}>"
+            code.extend(
+                f"""
+{ctype} {out_name} = as_array<{res.ctype}, {t.size}>({in_name});
+            """.split(
+                    "\n"
+                )
+            )
+        # we have to use c10::List for optional element. e.g., Tensor?[] -> c10::List<c10::optional<at::Tensor>>
+        elif isinstance(t.elem, OptionalType):
+            ctype = f"c10::List<{res.ctype}>"
+            code.extend(
+                f"""
+{ctype} {out_name};
+for (c10::IValue {elem_name}: {in_name}) {{
+    {connector.join(res.code)}
+    {out_name}.push_back({res.val_name});
+}}
+            """.split(
+                    "\n"
+                )
+            )
+        else:
+            # use ArrayRef as default.
+            ctype = f"at::ArrayRef<{res.ctype}>"
+            vec_name = arg_name + "_vec"
+            code.extend(
+                f"""
+std::vector<{res.ctype}> {vec_name};
+for (c10::IValue {elem_name}: {in_name}) {{
+    {connector.join(res.code)}
+    {vec_name}.push_back({res.val_name});
+}}
+{ctype} {out_name}({vec_name});
+            """.split(
+                    "\n"
+                )
+            )
+        return ArgumentCppCode(
+            code=code,
+            val_name=out_name,
+            ctype=ctype,
+        )
+    else:
+        raise Exception(f"Cannot handle type {t}. ival: {ival}, arg_name: {arg_name}")
+
+
+# Generate code to call C++ unboxed kernel with argument variable names
+def generate_unboxed_kernel_call(
+    f: NativeFunction, sig: CppSignature, arguments: Dict[str, ArgumentCppCode]
+) -> List[str]:
+    use_tensor_options = any(
+        isinstance(a.nctype.type, BaseCType) and a.nctype.type.type == tensorOptionsT
+        for a in sig.arguments()
+    )
+    arg_connector = ",\n\t"
+    # find dispatch native function namespace
+    if use_tensor_options:
+        namespace = "torch"
+    else:
+        namespace = "at"
+    # handle void return type
+    ret_type: CType = cpp.returns_type(f.func.returns)
+    if isinstance(ret_type, BaseCType) and ret_type.type == voidT:
+        ret_str = ""
+        push_str = ""
+    else:
+        ret_str = "auto result_ = "
+        push_str = """
+pack(stack, std::move(result_));
+        """
+    if Variant.method in f.variants:
+        self_arg = f.func.arguments.self_arg
+        assert self_arg is not None, "No self argument"
+        arg_list = arg_connector.join(
+            [
+                arguments[a.name].val_name
```

Comment:
I was trying to figure out why `arguments` needs to be a `Dict[str, ArgumentCppCode]` (and why the `ArgumentCppCode` class is necessary) and I think there are some existing abstractions in the codegen that should help to make this re-use patterns that the existing codegen uses.

This code is trying to call into a C++ API function (e.g. `at::add(at::Tensor& self, at::Tensor& other, at::Scalar& alpha);), and there are a bunch of local variables in scope that were created by unboxing the ivalues.

We have a `translate()` API that basically let's you do just that (lives and is documented here: https://github.com/pytorch/pytorch/blob/ce6e6812b1218363af91302bf64068a61c006278/tools/codegen/api/translate.py#L53).

It involves calling a function that has some target schema (in the above example, it's `Tensor& self`, `Tensor& other` and `Scalar& alpha`), while you have some context of local bindings that are in scope (those would be the temporaries that you constructed in the function from the IValues). `translate()` that returns the list of expressions that you need to pass into the function.

You can see an example of how it's used in a few places, e.g. [here](https://github.com/pytorch/pytorch/blob/ce6e6812b1218363af91302bf64068a61c006278/tools/codegen/dest/register_dispatch_key.py#L616).

Doing that should also force you to re-use more of the existing data model in codegen, instead of making new classes:

(1) You'll need to create a `List[Binding]` corresponding to the variables that are in scope when you're ready to call the function (example [here](https://github.com/pytorch/pytorch/blob/ce6e6812b1218363af91302bf64068a61c006278/tools/codegen/dest/register_dispatch_key.py#L581)).

To do (1), I think you'll want to have a function that operators on an individual `Argument`, and returns a tuple of:
- the code string that's used to unboxed that argument
- The `Binding` object that corresponds to the type of that unboxed argument, with the new (temporary variable) name that it was given.

There's an example piece of codegen that does that for the functionalization pass [here](https://github.com/pytorch/pytorch/blob/ce6e6812b1218363af91302bf64068a61c006278/tools/codegen/gen_functionalization_type.py#L43). If you look at where that function is used, you can see that the context it returns is fed directly into a translate call.

(2) the target bindings just correspond to the arguments in the faithful C++ API, since that's what you're calling. The `CppSignature` object represents the function that you're trying to call, so you can use it to get both the target bindings and the actual function call. Something like:
```
TORCH_API void {f.func.name.unambiguous_name()}(Stack & stack) {{
    ...
    at::{faithful_cpp_sig.name()}({', '.join(e.expr for e in translate(your_context, faithful_cpp_sig.arguments()))});
    ....
}}
```
Another advantage of doing it that way is that more of the "boilerplate" string manipulation can all live in the same place, so you don't have to go hunting around for different functions to piece together what the structure of the full codegen output looks like.

Reply:
Thanks for the detailed explanation. Let me give it a try!

Reply:
This is resolved.